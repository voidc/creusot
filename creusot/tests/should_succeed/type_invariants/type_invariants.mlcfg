
module TypeInvariants_WithInvariant_Type
  type t_withinvariant  =
    | C_WithInvariant
    
end
module TypeInvariants_Impl0_Invariant_Stub
  use TypeInvariants_WithInvariant_Type as TypeInvariants_WithInvariant_Type
  predicate invariant' [#"../type_invariants.rs" 9 4 9 30] (self : TypeInvariants_WithInvariant_Type.t_withinvariant)
end
module TypeInvariants_Impl0_Invariant_Interface
  use TypeInvariants_WithInvariant_Type as TypeInvariants_WithInvariant_Type
  predicate invariant' [#"../type_invariants.rs" 9 4 9 30] (self : TypeInvariants_WithInvariant_Type.t_withinvariant)
end
module TypeInvariants_Impl0_Invariant
  use TypeInvariants_WithInvariant_Type as TypeInvariants_WithInvariant_Type
  predicate invariant' [#"../type_invariants.rs" 9 4 9 30] (self : TypeInvariants_WithInvariant_Type.t_withinvariant) =
    [#"../type_invariants.rs" 10 8 10 12] true
  val invariant' [#"../type_invariants.rs" 9 4 9 30] (self : TypeInvariants_WithInvariant_Type.t_withinvariant) : bool
    ensures { result = invariant' self }
    
end
module TypeInvariants_Impl0_IsInhabited_Stub
  use TypeInvariants_WithInvariant_Type as TypeInvariants_WithInvariant_Type
  clone TypeInvariants_Impl0_Invariant_Stub as Invariant0
  function is_inhabited [#"../type_invariants.rs" 15 4 17 20] (_1' : ()) : ()
end
module TypeInvariants_Impl0_IsInhabited_Interface
  use TypeInvariants_WithInvariant_Type as TypeInvariants_WithInvariant_Type
  clone TypeInvariants_Impl0_Invariant_Stub as Invariant0
  function is_inhabited [#"../type_invariants.rs" 15 4 17 20] (_1' : ()) : ()
  axiom is_inhabited_spec : forall _1' : () . [#"../type_invariants.rs" 14 4 14 45] exists x : TypeInvariants_WithInvariant_Type.t_withinvariant . Invariant0.invariant' x /\ Invariant0.invariant' x
end
module TypeInvariants_Impl0_IsInhabited
  use TypeInvariants_WithInvariant_Type as TypeInvariants_WithInvariant_Type
  clone TypeInvariants_Impl0_Invariant_Stub as Invariant0
  function is_inhabited [#"../type_invariants.rs" 15 4 17 20] (_1' : ()) : () =
    [#"../type_invariants.rs" 13 4 13 10] ()
  val is_inhabited [#"../type_invariants.rs" 15 4 17 20] (_1' : ()) : ()
    ensures { result = is_inhabited _1' }
    
  axiom is_inhabited_spec : forall _1' : () . [#"../type_invariants.rs" 14 4 14 45] exists x : TypeInvariants_WithInvariant_Type.t_withinvariant . Invariant0.invariant' x /\ Invariant0.invariant' x
end
module TypeInvariants_Impl0_IsInhabited_Impl
  use TypeInvariants_WithInvariant_Type as TypeInvariants_WithInvariant_Type
  clone TypeInvariants_Impl0_Invariant as Invariant0
  let rec ghost function is_inhabited [#"../type_invariants.rs" 15 4 17 20] (_1' : ()) : ()
    ensures { [#"../type_invariants.rs" 14 4 14 45] exists x : TypeInvariants_WithInvariant_Type.t_withinvariant . Invariant0.invariant' x /\ Invariant0.invariant' x }
    
   = [@vc:do_not_keep_trace] [@vc:sp]
    [#"../type_invariants.rs" 13 4 13 10] ()
end
module TypeInvariants_Id_Interface
  use TypeInvariants_WithInvariant_Type as TypeInvariants_WithInvariant_Type
  clone TypeInvariants_Impl0_Invariant_Stub as Invariant0
  val id [#"../type_invariants.rs" 22 0 22 44] (x : TypeInvariants_WithInvariant_Type.t_withinvariant) : TypeInvariants_WithInvariant_Type.t_withinvariant
    requires {[#"../type_invariants.rs" 22 10 22 11] Invariant0.invariant' x}
    ensures { [#"../type_invariants.rs" 22 0 22 44] Invariant0.invariant' result }
    
end
module TypeInvariants_Id
  use TypeInvariants_WithInvariant_Type as TypeInvariants_WithInvariant_Type
  clone TypeInvariants_Impl0_Invariant as Invariant0
  clone TypeInvariants_Impl0_IsInhabited as IsInhabited0 with
    predicate Invariant0.invariant' = Invariant0.invariant',
    axiom .
  let rec cfg id [#"../type_invariants.rs" 22 0 22 44] [@cfg:stackify] [@cfg:subregion_analysis] (x : TypeInvariants_WithInvariant_Type.t_withinvariant) : TypeInvariants_WithInvariant_Type.t_withinvariant
    requires {[#"../type_invariants.rs" 22 10 22 11] Invariant0.invariant' x}
    ensures { [#"../type_invariants.rs" 22 0 22 44] Invariant0.invariant' result }
    
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : TypeInvariants_WithInvariant_Type.t_withinvariant;
  var x_1 : TypeInvariants_WithInvariant_Type.t_withinvariant;
  {
    x_1 <- x;
    goto BB0
  }
  BB0 {
    _0 <- x_1;
    x_1 <- any TypeInvariants_WithInvariant_Type.t_withinvariant;
    return _0
  }
  
end
module TypeInvariants_Impl0
  use TypeInvariants_WithInvariant_Type as TypeInvariants_WithInvariant_Type
  clone TypeInvariants_Impl0_Invariant as Invariant0
  goal is_inhabited_refn : [#"../type_invariants.rs" 15 4 17 20] forall result : () . (exists x : TypeInvariants_WithInvariant_Type.t_withinvariant . Invariant0.invariant' x /\ Invariant0.invariant' x) -> (exists x : TypeInvariants_WithInvariant_Type.t_withinvariant . Invariant0.invariant' x)
end
