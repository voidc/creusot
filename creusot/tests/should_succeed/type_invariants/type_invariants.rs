extern crate creusot_contracts;
use creusot_contracts::*;

pub struct WithInvariant;

impl WithInvariant {
    #[predicate]
    #[creusot::type_invariant]
    fn invariant(self) -> bool {
        true
    }
}

pub fn id(x: WithInvariant) -> WithInvariant {
    x
}