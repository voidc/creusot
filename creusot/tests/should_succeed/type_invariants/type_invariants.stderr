warning: associated function `invariant` is never used
 --> type_invariants.rs:9:8
  |
9 |     fn invariant(self) -> bool {
  |        ^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: 1 warning emitted

